loader/src/ui/nodes/ProgressBar.cpp

```diff
@@ -0,0 +1,178 @@
+#include <cocos2d.h>
+#include <Geode/ui/ProgressBar.hpp>
+
+using namespace geode::prelude;
+
+class ProgressBar::Impl final {
+public:
+    // Progress bar outline
+    Ref<CCSprite> progressBar = nullptr;
+    // Progress bar fill
+    CCSprite* progressBarFill = nullptr;
+    // The text label displaying the percentage
+    Ref<CCLabelBMFont> progressPercentLabel = nullptr;
+
+    // Current progress bar fill percentage ranging from 0 to 100
+    float progress = 0.0f;
+
+    // Style of the progress bar
+    ProgressBarStyle style = ProgressBarStyle::Level;
+    // Current color of the filled progress bar
+    ccColor3B progressBarFillColor = { 255, 255, 255 };
+    // Whether to show the label showing the percentage of the current progress
+    bool showProgressPercentLabel = false;
+
+    // Max width for the progress fill bar node
+    float progressBarFillMaxWidth = 0.0f;
+    // Max height for the progress fill bar node
+    float progressBarFillMaxHeight = 0.0f;
+};
+
+ProgressBar::ProgressBar() {
+    m_impl = std::make_unique<Impl>();
+};
+
+ProgressBar::~ProgressBar() {};
+
+void ProgressBar::customSetup() {
+    switch (m_impl->style) {
+    case ProgressBarStyle::Level:
+        m_impl->progressBar = CCSprite::create("slidergroove2.png");
+        m_impl->progressBar->setID("progress-bar");
+        m_impl->progressBar->setAnchorPoint({ 0.5, 0.5 });
+        m_impl->progressBar->setPosition({ m_impl->progressBar->getScaledContentWidth() / 2.0f, m_impl->progressBar->getScaledContentHeight() / 2.0f });
+        m_impl->progressBar->setZOrder(1);
+
+        m_impl->progressBarFill = CCSprite::create("sliderBar2.png");
+        m_impl->progressBarFill->setID("progress-bar-fill");
+        m_impl->progressBarFill->setAnchorPoint({ 0, 0.5 });
+        m_impl->progressBarFill->setPosition({ 2.0f, m_impl->progressBar->getScaledContentHeight() / 2.0f });
+        m_impl->progressBarFill->setColor(m_impl->progressBarFillColor);
+        m_impl->progressBarFill->setZOrder(-1);
+
+        m_impl->progressBarFillMaxWidth = m_impl->progressBar->getScaledContentWidth() - 4.0f;
+        m_impl->progressBarFillMaxHeight = m_impl->progressBarFill->getScaledContentHeight() - 0.5f;
+
+        m_impl->progressPercentLabel = CCLabelBMFont::create("0%", "bigFont.fnt");
+        m_impl->progressPercentLabel->setID("progress-percent-label");
+        m_impl->progressPercentLabel->setScale(0.5f);
+        m_impl->progressPercentLabel->setAnchorPoint({ 0, 0.5 });
+        m_impl->progressPercentLabel->setPosition({ m_impl->progressBar->getScaledContentWidth() + 2.5f, m_impl->progressBar->getScaledContentHeight() / 2.0f });
+        m_impl->progressPercentLabel->setAlignment(CCTextAlignment::kCCTextAlignmentLeft);
+        m_impl->progressPercentLabel->setVisible(m_impl->showProgressPercentLabel);
+        m_impl->progressPercentLabel->setZOrder(1);
```

Comment:
This is just a styling thing but I would indent this whole thing one tab and then surround the block with braces, like so:
```cpp
switch (x) {
    case y: {
        // code here...
    } break;
}
```